ecommerce/extensions/payment/processors/liqpay.py

```diff
@@ -0,0 +1,221 @@
+            request (Request, optional): A Request object which can be used to construct an absolute URL in
+                cases where one is required.
+            use_client_side_checkout (bool, optional): Determines if client-side checkout should be used.
+            **kwargs: Additional parameters.
+
+        Returns:
+            dict: Payment processor-specific parameters required to complete a transaction. At a minimum,
+                this dict must include a `payment_page_url` indicating the location of the processor's
+                hosted payment page.
+        """
+
+        params = self._generate_parameters(basket)
+        parameters = {
+            'payment_page_url': self.payment_page_url,
+            'signature': self.make_signature(params),
+            'data': b64encode(json.dumps(params).encode("utf-8")).decode("ascii"),
+        }
+        return parameters
+
+    def _generate_parameters(self, basket):
+        """ Generates the parameters dict.
+
+        A signature is NOT included in the parameters.
+
+        Arguments:
+            basket (Basket): Basket from which the pricing and item details are pulled.
+
+        Returns:
+            dict: Dictionary containing the payment parameters that should be sent to Liqpay.
+        """
+
+        result_url = get_receipt_page_url(
+            order_number=basket.order_number,
+            site_configuration=basket.site.siteconfiguration
+        )
+
+        parameters = {
+            'public_key': self.public_key,
+            'action': 'pay',
+            'amount': str(basket.total_incl_tax),
+            # available default currencies in liqpay are USD, EUR, RUB, UAH, BYN, KZT
+            'currency': basket.currency,
+            'description': basket.order_number,
+            'order_id': basket.order_number,
+            'version': self.version,
+            'sandbox': self.sandbox_mode,
+            'server_url': urljoin(get_ecommerce_url(), reverse('liqpay:callback')),
+            'result_url': result_url,
+        }
+        return parameters
+
+    def handle_processor_response(self, response, basket=None):
+        """
+        Handle a response from the payment processor.
+
+        This method creates PaymentEvents and Sources for successful payments.
+
+        Arguments:
+            response (dict): Dictionary of parameters received from the payment processor
+
+        Keyword Arguments:
+            basket (Basket): Basket whose contents have been purchased via the payment processor
+
+        Returns:
+            HandledProcessorResponse
+        """
+
+        # Validate the signature (indicating potential tampering)
+        data = response.get('data')
+        sign_string = str(self.private_key) + str(data) + str(self.private_key)
+        sign = b64encode(hashlib.sha1(sign_string.encode("utf-8")).digest()).decode("ascii")
+        if sign != response.get('signature'):
+            raise InvalidSignatureError
```

Comment:
Is it make sense to do a manful message to tell what exactly wrong with the given signature?

Reply:
done